### PR TITLE
Printing unfocussed goals and toward a pg plugin.

### PR DESCRIPTION
**Pull request by Matafou**
I had several wishes to see unfocussed goals.

I also added two other options concerning goal printing, which has the effect of removing the uses to emacs_str in printer.ml. Hopefully this way one day pg support will become a plugin.

### CODE REVIEW

**Comment by Matafou** [comment]
This will need to be propagated in plugins that replace pr_subgoals because this function must takes one more argument now (the list of unfocussed goals).

Feel free to suggest a better way to do this.

**Comment by ejgallego** [comment]
The general idea is good, however note #455 and #390 , this will likely have to be merged after those land.

**Comment by maximedenes** [comment]
Hi Pierre,

Thanks for contributing! Can you explain a bit more the use cases? A naive question: if one needs to see the unfocused goals, why focus other goals in the first place?

**Comment by jonleivent** [comment]
@maximedenes : one reason to want to see unfocused goals is to see how shared evars impact them.  Those goals may have become unfocused before the user becomes aware that a specific shared evar needs to be filled in to solve some focused goal.

**Comment by Matafou** [comment]
Focusing happens for example when you apply a bullet, but I don't think bulleting should necessarily hide the other bullet's goals. Actually it may help to see them.

Moreover focusing is also made to change the range of all: goal specifier. Again this does not mean that one wants the goals not in the range of all: should be hidden.

Of course unfocused goals should be tagged as such.

**Comment by maximedenes** [comment]
I need to think about @jonleivent's answer. But the bullet story doesn't convince me. Isn't it rather a sign that bullets should not focus? As for the `all:` specifier, I'd need to see a concrete use case to see how focusing is done and why, in that specific case.

I'm just trying to make sure this is not an XY problem.

**Comment by mattam82** [comment]
I think bullets should stay focusing, and one does not necessarily always want to see the unfocused ones (although I can see the point when you share dependencies with them).

**Comment by maximedenes** [comment]
I'm fine with anything, and we are certainly not going to change that now, but out of curiosity, what is the point of having bullets focus? I understand why we want them to be strict (i.e. get an error if we forget one, etc.). But focusing in this case has always seemed to be rather annoying to me in practice (precisely, I wanted to see the other goals).

**Comment by Zimmi48** [comment]
@maximedenes Do you have an alternative proposal for bullets? First of all, I don't see how they could be made strict without focusing. Secondly, I feel like the focusing is part of the message that they are sending to the user: I'm now going to be solving the first sub-goal, I'm now going to be solving the second sub-goal, etc. Without focusing, you cannot guarantee that you are not actually solving the second sub-goal in the first bullet.

In fact, I really feel like having bullets focus is the right choice but choosing to mask unfocused sub-goals by default was the real mistake.

**Comment by maximedenes** [comment]
> Without focusing, you cannot guarantee that you are not actually solving the second sub-goal in the first bullet.

But focusing has nothing to do with this guarantee either:

```
Definition is_zero n := n = O.

Lemma foo : True -> is_zero O.
Proof.
reflexivity.
Qed.

Goal exists n, is_zero n.
unshelve refine (ex_intro _ _ _); cycle 1.
+ apply foo. (* solves the second, unfocused goal *)
```

**Comment by Matafou** [comment]
IMO bullets are a just a convenient way to use focusing with a tree structure (not necessarily matching the tactic structure by the way).

Here is a typical example of the use of focusing bullets:

Lemma foo: forall x y:nat, P x y.
Proof.
  induction x.
  - destruct y.
    all: simpl; omega. (* only acts on the 2 focused subgoals*)
  - ...

Actually bullets themselves need a focusing mechanism when you nest them anyway, since bullets apply on currently focused subgoals.

Lemma foo: forall x y:nat, P x y.
Proof.
  induction x.
  - destruct y.
    + ...  (* the bullet + is applied on the two current focused subgoals *)
    + ...
  - ...

**Comment by Zimmi48** [comment]
Your option does not seem to have any effect in CoqIDE. Can you make it so that the unfocused goals are shown there too?
I would even argue to make printing unfocused goals the default.

Another remark is that in coqtop, the option does not just show the unfocused goals. It also shows some additional information on each displayed goals (e.g. `subgoal 1 (?Goal)`).

File: printing/printer.ml
Comment (herbelin) [inline printing/printer.ml]: At first view, the parser will not be able to parse the space. You should do ["Printing";"Goal";"Tags"].

(Unless the space is intended, so that the option can only be used over a dedicated communication channel which accepts spaces in the name of the option?)

File: printing/printer.ml
Comment (herbelin) [inline printing/printer.ml]: Idem

File: printing/printer.ml
Comment (herbelin) [inline printing/printer.ml]: The usual issue: do we want the command to be possibly part of a document or not, and hence synchronized with backtracking?

The usage seems that the `Printing` commands are `sync`, so by consistency, until a protocol is found to deal with options usable both as a command and from a menu, I would lean towards making the option `sync`.

File: printing/printer.ml
Comment (herbelin) [inline printing/printer.ml]: Both _focussed_ and _focused_ are apparently correct English. The code and the reference manual use _focused_. This would be easier to remember if it were always the same spelling, hence with a single _s_ to continue as it is.

File: printing/printer.ml
Comment (Matafou) [inline printing/printer.ml]: OK for consistency.

File: printing/printer.ml
Comment (Matafou) [inline printing/printer.ml]: No, that was a typo.

File: printing/printer.ml
Comment (Matafou) [inline printing/printer.ml]: OK for focused.

**Comment by Matafou** [comment]
@Zimmi48 good point. I will have a look at it.

**Comment by ejgallego** [comment]
@Matafou @Zimmi48 for CoqIDE you may want to tweak `ide/wg_proofview.ml:display` I think, but that would IMHO be a different PR.

File: printing/printer.ml
Comment (Matafou) [inline printing/printer.ml]: But we really need to have both ways of setting options: synchronized and unsynchronized.

**Comment by Matafou** [comment]
All these suggestions implemented, except the coqide part.

**Comment by Matafou** [comment]
Where should I put the options to be able to check them from ide/wg_proofview.ml?

**Comment by herbelin** [comment]
> But we really need to have both ways of setting options: synchronized and unsynchronized.

Yes, I agree. I made a proposal to deal with that at https://coq.inria.fr/bugs/show_bug.cgi?id=5229. What do you think? Or do you have another model in mind?

**Comment by herbelin** [comment]
@maximedenes: But in which sense of the word "fragile" do you say it was fragile before!!!

**Comment by Matafou** [comment]
I think I got it right in coqide (behaves like other "known" printing options. We can probably merge now if tests are ok.

@ejgallego I agree that option should have both ways of being set. I just proposed a variant of you proposal in the bug report.

**Comment by Zimmi48** [comment]
@Matafou I think you wanted to say @herbelin instead of @ejgallego here, no?

**Comment by Matafou** [comment]
indeed @Zimmi48. Thanks.

**Comment by Matafou** [comment]
Some of the tests in test-suite fail, is it specific to this PR?

**Comment by ejgallego** [comment]
@Matafou yes, you have likely modified the output of some commands.

**Comment by ejgallego** [comment]
Test suite complains:
```
    output/Intuition.v...Error! (unexpected output)
    output/Naming.v...Error! (unexpected output)
    output/Quote.v...Error! (unexpected output)
    output/set.v...Error! (unexpected output)
    output/simpl.v...Error! (unexpected output)
    output/subst.v...Error! (unexpected output)
    output/unifconstraints.v...Error! (unexpected output)
```

**Comment by Matafou** [comment]
OK only fiat-crypto fails now. I guess it is mergeable.

@maximedenes do you have still doubts about bullets vs focusing?

**Comment by maximedenes** [comment]
> @maximedenes do you have still doubts about bullets vs focusing?

Yes, but that doesn't prevent us from merging this PR :)

**Comment by maximedenes** [comment]
I relaunched the fiat crypto test, we'll see.

**Comment by JasonGross** [comment]
Does this change the default printout in PG?  I frequently focus goals *specifically* to hide the other goals, sometimes because Coq is very slow at printing.

**Comment by Zimmi48** [comment]
A few remarks:
- This makes printing unfocused sub-goals the default in CoqIDE. I like it (and even suggested it) but some others like @JasonGross may disagree.
- This does not make it the default in coqtop / PG, which is kind of inconsistent I would think.
- In PG I can see a small unrelated change in the printing:

```
2 subgoals, subgoal 1 (?Goal)
  
  ============================
  True

subgoal 2 (?Goal0) is:
 True
```

instead of

```
2 subgoals, subgoal 1 (ID 3)
  
  ============================
  True

subgoal 2 (ID 4) is:
 True
```

**Comment by Matafou** [comment]
@JasonGross @Zimmi48 You are right I should not change the default behavior. I will fix that.

@Zimmi48 I don't see ?Goal unless "Printing Goal Names" is set (which is an experimental feature I understand). And to see ID you need to set the Printing Goal Tags option. this is for proof tree mode in pg. 
did you pull last version? There was a bug with these options that I fixed.

**Comment by Zimmi48** [comment]
> did you pull last version?

Yes, I did. I'm surprised that you don't see the same things as me. My version of Proof General is 4.4pre and I'm also using Company-Coq. Could it be due to that?

**Comment by Matafou** [comment]
Ha yes, I see it now. Fixed.

**Review by Zimmi48 — APPROVED** [review]

**Comment by psteckler** [comment]
PG/xml receives all goals, focused/background/shelved from coqtop. It prints full details only for the focused goal, but that could be changed. The old PG received goals already formatted from coqtop, while PG/xml receives the goal data in XML trees, and does its own formatting.

**Comment by Zimmi48** [comment]
@psteckler Do you have a way to add an option to PG/xml to show unfocused goals, similar to what Pierre did on CoqIDE? That would be great to have!

**Comment by psteckler** [comment]
I can add such an option. Besides unfocused goals, PG/xml receives information about shelved and abandoned goals, if any. So the option could be to show full details for all of those goals.

Do you want to add a feature request at https://github.com/psteckler/ProofGeneral/issues ?

**Comment by maximedenes** [comment]
@Matafou I think you wanted to rebase this on top of `trunk` instead of merging `trunk`, that usually results in spaghetti history.

**Comment by Matafou** [comment]
yeah I will do that, Never di this on a PR though, will it maintian the pr
page and links?
P.



2017-04-24 15:25 GMT+02:00 Maxime Dénès <notifications@github.com>:

**Comment by Zimmi48** [comment]
@Matafou It shouldn't be a problem. Other people do that all the time. Of course, you will need to force-push.

**Comment by Matafou** [comment]
I just squashed all those commits into one.

**Comment by Matafou** [comment]
@maximedenes Concerning your doubts about bullets and focusing I think we should clarify things:

"Focus" is a very old command that dates back to the terminal-toplevel time where seeing other goals meant scrolling your current goal (printed atop of the others) out of view. So Focus at that time meant "hide the other goals".

But hiding is pointless most of the time now because IDEs know how to show the current goal even when the other ones are output by coqtop.

Bullets and curly brackets are now used for two new purposes:
1) define the scope of goal selectors.
2) help the user to structure the script and maintain this structure.

The purpose 1) is not very used currently but may become common.
The purpose 2) proved very useful.

 Note: the interaction between 1) and 2) is not completely clear to me.

The confusion comes from the fact that bullets were implemented by reusing the "Focus" mechanism. So as a side effect the goals not under the current bullet got hidden. But printing them (making clear that they are not currently operable.

**Comment by maximedenes** [comment]
> The confusion comes from the fact that bullets were implemented by reusing the "Focus" mechanism.

Yes, but my point was: shouldn't this be fixed instead?

Note that I don't think it should slow down the integration of this PR, but I wanted to understand if it was the proper fix, or just a workaround for the flawed design.

**Comment by Matafou** [comment]
Actually @aspiwack  implemented this new meaning of "focussing" in a very reasonable way IMHO. It is just the fact that Focus changed its meaning at that moment.

**Comment by Zimmi48** [comment]
> Yes, but my point was: shouldn't this be fixed instead?

Would you completely change the implementation of bullets, while keeping the "Focus" mechanism?
Since the original purpose of `Focus` is not needed these days, isn't this PR fixing the problem (if we just make printing unfocused goals the default)?
Or are you worried about the interaction between points 1 and 2 of @Matafou's comment?

**Comment by JasonGross** [comment]
> But hiding is pointless most of the time now 

Hiding, implemented on the Coq side, is an absolutely essential feature for some of my developments; when it takes upwards of ten seconds to print one subgoal, having Coq always send all goals would massively slow development.  I would like Coq-side goal-hiding to stick around at least until Coq is better at printing large terms, or at least better at printing terms with large implicit or otherwise hidden subterms.

**Comment by Zimmi48** [comment]
@JasonGross Yes, I'm sorry to have talked again of changing the default printing state while you had already communicated your opposition to this earlier in this thread.
Anyways, as soon as the option to show unfocused goals exists, people can just turn it on if they wish.

**Comment by JasonGross** [comment]
@Zimmi48 no worries.  And I actually have only a weak objection to the default changing (since I can just change it back in my resource file), as long as the option to hide goals doesn't disappear entirely.  In the long term, I would hope that some of the work @ejgallego is doing on the ast and ide side of things will somehow fix the problem that Coq spends a lot of time detyping(?) subterms that are later hidden.

**Comment by ejgallego** [comment]
@JasonGross, SerAPI has already support for selecting a subset of goals and even hypothesis for display, this avoids any work at all on the things that should not be displayed. What is missing now is UI support for it and in particular a the proper query syntax, but that should be easily doable in PG-SerAPI.

Of course, improving printing speed is a high-priority goal but not the primary solution for your use case.

**Comment by JasonGross** [comment]
@ejgallego that doesn't handle this case of goal printing being slow, does it? (Note: this particular value of `x` is untested; you may have to fiddle with it to get something that's show enough.)
```coq
Notation hide := (_ = _).
Definition x := List.repeat 5 Z.div_eucl.
Goal x = x.
  vm_compute.
  Show. (* I think this is slow, even though it only eventually prints [hide] *)
  Show.
```

**Comment by ejgallego** [comment]
@JasonGross I can see ways of making that work fast but IMHO much work is needed first including Phase 2 of CEP012

**Comment by herbelin** [comment]
Note that there is a known inefficiency in turning De Bruijn indices to names in `detyping.ml` (so before notations are considered). @aspiwack had a look at it once, iirc. It is something like n^3 in the depth of binders.

File: ide/wg_ProofView.ml
Comment (maximedenes) [inline ide/wg_ProofView.ml]: Please make this a label.

File: ide/wg_ProofView.ml
Comment (maximedenes) [inline ide/wg_ProofView.ml]: `shownum` should also be a label

File: printing/printer.ml
Comment (maximedenes) [inline printing/printer.ml]: `str ""` should be replaced by `mt ()` I believe

File: ide/wg_ProofView.ml
Comment (Matafou) [inline ide/wg_ProofView.ml]: an optional one?

**Comment by Matafou** [comment]
@maximedenes is it what you had in mind?

I like labels but until now I refrained myself putting labels in coq code. Is there some policy about that?

**Comment by maximedenes** [comment]
@Matafou Yes, thanks! I believe labels are encouraged when the types and function name are not informative enough :)

**Comment by maximedenes** [comment]
Testing with `-warn-errors`, I realized that `delayed_emacs_cmd` is now dead code. Could you remove it? I'll merge this PR right after. Thanks!

**Comment by aspiwack** [comment]
@maximedenes aside: it is generally considered good practice to have `-warn-errors` turned on in CI.

**Comment by maximedenes** [comment]
@aspiwack Yes, I couldn't follow the discussion between @ejgallego and @SkySkimmer, but I was surprised that this was not the outcome.

**Comment by ejgallego** [comment]
> it is generally considered good practice to have -warn-errors turned on in CI.

IMHO if `-warn-errors` is not the default setup for developers then CI becomes a large waste of time. I am sorry but I do not want to waste time figuring out why things are OK in my setup and not in the CI.

**Comment by ejgallego** [comment]
In other words, we badly need to reduce the number of configurations we need to test.

**Comment by maximedenes** [comment]
@ejgallego I don't have a strong opinion here, but why does it need to be the default? As long as *your* setup matches Travis, it's fine, isn't it?

In fact, today's defaut is not to build the doc, which also doesn't match Travis, since one of the jobs does.

**Comment by ejgallego** [comment]
> today's default is not to build the doc,

Which in fact has led to quite a few people to submit PR with the docs broken, as they didn't test them locally, and a large waste of time. If we enable `-warn-on-error` only on the CI, I predict half of the PR will be broken due to warnings, and we will all incur in a large waste of time.

It makes no sense to have a different default for the required CI and the default dev build setup.

**Comment by SkySkimmer** [comment]
If @maximedenes keeps using warn-error and asking people to fix it that's even slower than travis ;)

**Comment by maximedenes** [comment]
@SkySkimmer Indeed, that was my point, implicitly. Also, it is a bit tiring :)

But I'm also fine if we decide to live with warnings and dead code.

**Comment by maximedenes** [comment]
> If we enable -warn-on-error only on the CI, I predict half of the PR will be broken due to warnings, and we will all incur in a large waste of time.

I think it will happen once per dev, but then, they will update their settings. But again, I'm fine with any solution, as long as what I need to do is clear.

**Comment by ejgallego** [comment]
So this is the point I can't follow, if we are going to require every dev to enable `-warn-on-error`, why is just not enabled by default?

**Comment by maximedenes** [comment]
`-warn-on-error` is a good default setting for developers, but not for distribution (people who want to clone, compile and use Coq.

**Comment by ejgallego** [comment]
Then, enable it for developers (version number `+gitXXX`) and disable it for the rest.

**Comment by Zimmi48** [comment]
Sometimes, people forgot to build the doc. Sometimes, people even forgot to build the test-suite.
Can we provide a `make` target that will turn the `warn-on-error` option on, build the doc and the test-suite?

**Comment by SkySkimmer** [comment]
This should probably be taken to coqdev or ceps or something.

**Comment by ejgallego** [comment]
Indeed the summary of our discussion about warnings was "lets merge this with very conservative settings, and delay discussion to a WG"

**Comment by maximedenes** [comment]
> Would you people leave my PR alone and go to gitter!! :-)

Sorry, I plead guilty :)

**Comment by Matafou** [comment]
There is a warning in tactics/eqdecide.ml that does not seem due to me. Fixing it in this PR would be out of scope I suppose.

**Comment by maximedenes** [comment]
Merged, thanks!

**Comment by Zimmi48** [comment]
@Matafou As this ever been documented? If not, it should be.

**Comment by Matafou** [comment]
You are right. Were should I propose a PR? in master or 8.7?

P.

2017-08-16 15:37 GMT+02:00 Théo Zimmermann <notifications@github.com>:

**Comment by Zimmi48** [comment]
On master please.

**Comment by Matafou** [comment]
Done.
P.

2017-08-16 16:30 GMT+02:00 Théo Zimmermann <notifications@github.com>:

**Comment by Zimmi48** [comment]
That's impressive!

**Pull request merged**
